src/popup/PopupLayer.tsx

```diff
@@ -0,0 +1,103 @@
+import * as React from 'react';
+import noop from 'lodash/noop';
+import PopupReply from '../components/Popups/PopupReply';
+import { CreateArg as HighlightCreateArg } from '../highlight/actions';
+import { CreateArg as RegionCreateArg } from '../region/actions';
+import { CreatorItem, CreatorStatus, isCreatorStagedHighlight, isCreatorStagedRegion } from '../store';
+import { PopupReference } from '../components/Popups/Popper';
+import { Shape } from '../@types';
+import './PopupLayer.scss';
+
+type Props = {
+    createHighlight?: (arg: HighlightCreateArg) => void;
+    createRegion?: (arg: RegionCreateArg) => void;
+    isCreating: boolean;
+    isPromoting: boolean;
+    location: number;
+    message: string;
+    popupReference?: Shape;
+    resetCreator: () => void;
+    setMessage: (message: string) => void;
+    staged?: CreatorItem | null;
+    status: CreatorStatus;
+};
+
+const PopupLayer = (props: Props): JSX.Element | null => {
+    const {
+        createHighlight = noop,
+        createRegion = noop,
+        isCreating = false,
+        isPromoting = false,
+        message,
+        popupReference,
+        resetCreator,
+        setMessage,
+        staged,
+        status,
+    } = props;
+
+    const [reference, setReference] = React.useState<PopupReference | null>(null);
```

Comment:
Is the extra state here needed or can we derive and pass the `virtualElement` directly in render?

Reply:
I had originally done just that, but I think the fact that `virtualElement` was a new object each render caused some wonky behavior in IE11